File: ab-cloud-verification/javascript/ab_cloud_verify.js
// ==============================================================================
// AB-Cloud Verification Suite — JavaScript/Node.js (Bilingual: EN/RU)
// ==============================================================================
// Verifies three key objections against Riemann zeta zero data:
//   Objection 1: b(N) convergence  (Gram-point deviation via Lambert W)
//   Objection 2: GUE spacing KS test
//   Objection 3: Large-T decay slope ≈ -0.5
//
// Node.js 18+ required. Uses fs, readline, zlib (for .gz streams).
// ==============================================================================

'use strict';

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const zlib = require('zlib');

// ==============================================================================
// Bilingual message tables
// ==============================================================================

const MSG = {
  en: {
    header:       'AB-Cloud Verification Suite — JavaScript/Node.js',
    separator:    '──────────────────────────────────────────────────────',
    loading:      'Loading zeros from: %s',
    loaded:       'Loaded %d zeros from %s',
    obj1_title:   'Objection 1: b(N) Convergence',
    obj1_desc:    'b(N) = (1/N) * Σ|γ_k - γ̃_k|, Gram points via Lambert W',
    obj1_n:       'N',
    obj1_bN:      'b(N)',
    obj1_status:  'Status',
    obj1_converge:'CONVERGING — b(N) → 0 supports AB-Cloud',
    obj1_stable:  'STABLE — b(N) near zero, AB-Cloud consistent',
    obj1_diverge: 'DIVERGING — b(N) not → 0, objection upheld',
    obj2_title:   'Objection 2: GUE Spacing KS Test',
    obj2_desc:    's_k = (γ_{k+1}-γ_k)·log(γ_k/2π)/(2π), vs p(s)=(πs/2)·exp(-πs²/4)',
    obj2_stat:    'D-statistic',
    obj2_pval:    'p-value',
    obj2_result:  'Result',
    obj2_pass:    'PASS — GUE spacing confirmed (p > 0.05)',
    obj2_fail:    'FAIL — GUE spacing rejected (p ≤ 0.05)',
    obj3_title:   'Objection 3: Large-T Decay Slope',
    obj3_desc:    'Linear regression of log|γ_k - γ̃_k| vs log(γ_k), expect slope ≈ -0.5',
    obj3_slope:   'Slope',
    obj3_stderr:  'Std Error',
    obj3_target:  'Target',
    obj3_result:  'Result',
    obj3_pass:    'PASS — Slope ≈ -0.5, AB-Cloud decay confirmed',
    obj3_fail:    'FAIL — Slope deviates from -0.5',
    no_data:      'ERROR: No zeros loaded. Check data directory.',
    done:         'Verification complete.',
  },
  ru: {
    header:       'Комплекс проверки AB-Cloud — JavaScript/Node.js',
    separator:    '──────────────────────────────────────────────────────',
    loading:      'Загрузка нулей из: %s',
    loaded:       'Загружено %d нулей из %s',
    obj1_title:   'Возражение 1: Сходимость b(N)',
    obj1_desc:    'b(N) = (1/N) * Σ|γ_k - γ̃_k|, точки Грама через W Ламберта',
    obj1_n:       'N',
    obj1_bN:      'b(N)',
    obj1_status:  'Статус',
    obj1_converge:'СХОДИТСЯ — b(N) → 0 подтверждает AB-Cloud',
    obj1_stable:  'СТАБИЛЬНО — b(N) ≈ 0, AB-Cloud согласуется',
    obj1_diverge: 'РАСХОДИТСЯ — b(N) ↛ 0, возражение подтверждено',
    obj2_title:   'Возражение 2: KS-тест интервалов GUE',
    obj2_desc:    's_k = (γ_{k+1}-γ_k)·log(γ_k/2π)/(2π), сравн. с p(s)=(πs/2)·exp(-πs²/4)',
    obj2_stat:    'D-статистика',
    obj2_pval:    'p-значение',
    obj2_result:  'Результат',
    obj2_pass:    'ПРОЙДЕНО — интервалы GUE подтверждены (p > 0.05)',
    obj2_fail:    'НЕ ПРОЙДЕНО — интервалы GUE отклонены (p ≤ 0.05)',
    obj3_title:   'Возражение 3: Наклон убывания при больших T',
    obj3_desc:    'Регрессия log|γ_k - γ̃_k| от log(γ_k), ожид. наклон ≈ -0.5',
    obj3_slope:   'Наклон',
    obj3_stderr:  'Стд. ошибка',
    obj3_target:  'Цель',
    obj3_result:  'Результат',
    obj3_pass:    'ПРОЙДЕНО — Наклон ≈ -0.5, убывание AB-Cloud подтверждено',
    obj3_fail:    'НЕ ПРОЙДЕНО — Наклон отклоняется от -0.5',
    no_data:      'ОШИБКА: Нули не загружены. Проверьте каталог данных.',
    done:         'Проверка завершена.',
  },
};

// ==============================================================================
// Lambert W (principal branch) — Halley's method
// ==============================================================================

function lambertW0(x) {
  if (x === 0) return 0;
  let w = x > 1 ? Math.log(x) - Math.log(Math.log(x)) : x;
  for (let iter = 0; iter < 50; iter++) {
    const ew  = Math.exp(w);
    const f   = w * ew - x;
    const fp  = ew * (1 + w);
    const fpp = ew * (2 + w);
    w -= (2 * f * fp) / (2 * fp * fp - f * fpp);
    if (Math.abs(f) < 1e-12 * Math.abs(x + 1)) break;
  }
  return w;
}

// ==============================================================================
// Gram point via Lambert W with Newton refinement
// ==============================================================================

function gramPoint(n) {
  if (n <= 0) return 0;
  const TWO_PI = 2 * Math.PI;
  let g = TWO_PI * n / lambertW0(n / Math.E);
  // Newton refinement using exact θ(t)
  for (let iter = 0; iter < 3; iter++) {
    const theta  = 0.5 * g * Math.log(g / TWO_PI) - 0.5 * g - Math.PI / 8;
    const dtheta = 0.5 * Math.log(g / TWO_PI);
    g += (Math.PI * n - theta) / dtheta;
  }
  return g;
}

function gramPointsVec(nVec) {
  return nVec.map(n => gramPoint(n));
}

// ==============================================================================
// GUE spacing CDF (Wigner surmise)
// ==============================================================================

function gueCDF(s) {
  return 1 - Math.exp(-Math.PI * s * s / 4);
}

// ==============================================================================
// Load zeros from data files (async, streaming)
// ==============================================================================

const FILE_MAP = {
  zeta_zeros_50000:     'zeta_zeros_50000.txt',
  zeta_zeros_500k:      'zeta_zeros_500k_odlyzko.txt',
  zeta_zeros_2M:        'zeta_zeros_2M_odlyzko.txt',
  zeta_zeros_highT:     'zeta_zeros_highT_blocks.txt',
  zeros6:               'zeros6.txt',
  zeta_zeros_50000_csv: 'zeta_zeros_50000.csv',
};

async function loadZeros(dataDir, count = 0, source = 'auto') {
  let selectedFile = null;
  let selectedName = null;

  if (source !== 'auto' && FILE_MAP[source]) {
    selectedFile = path.join(dataDir, FILE_MAP[source]);
    selectedName = source;
  } else {
    // Auto-select: pick first available, prefer small file for small counts
    const order = Object.keys(FILE_MAP);
    for (const nm of order) {
      const fp = path.join(dataDir, FILE_MAP[nm]);
      if (fs.existsSync(fp)) {
        selectedFile = fp;
        selectedName = nm;
        if (count > 0 && count <= 50000 && nm === 'zeta_zeros_50000') break;
      }
    }
  }

  if (!selectedFile || !fs.existsSync(selectedFile)) {
    throw new Error(`No data file found in: ${dataDir}`);
  }

  console.log(`Loading zeros from: ${selectedFile}`);

  const zeros = [];
  const ext = path.extname(selectedFile).toLowerCase();
  const isGz = selectedFile.endsWith('.gz');
  const isCsv = ext === '.csv';

  // Create read stream (handle .gz via zlib)
  let stream = fs.createReadStream(selectedFile, { encoding: 'utf8' });
  if (isGz) {
    stream = fs.createReadStream(selectedFile).pipe(zlib.createGunzip()).setEncoding('utf8');
  }

  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

  let lineNumber = 0;
  for await (const line of rl) {
    lineNumber++;
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    if (isCsv) {
      // CSV: split by comma, take first column
      const parts = trimmed.split(',');
      const val = parseFloat(parts[0]);
      if (!isNaN(val) && isFinite(val)) zeros.push(val);
    } else {
      // TXT: parse all floats on the line (handles whitespace-prefixed)
      const tokens = trimmed.split(/\s+/).filter(t => t.length > 0);
      for (const tok of tokens) {
        const val = parseFloat(tok);
        if (!isNaN(val) && isFinite(val)) zeros.push(val);
      }
    }

    // Early exit if count specified
    if (count > 0 && zeros.length >= count) {
      rl.close();
      break;
    }
  }

  // Trim to exact count
  if (count > 0 && zeros.length > count) zeros.length = count;

  console.log(`Loaded ${zeros.length} zeros from ${path.basename(selectedFile)}`);
  return zeros;
}

// ==============================================================================
// Objection 1: b(N) Convergence
// ==============================================================================

function objection1(zeros, M) {
  const N = zeros.length;
  console.log(`\n${M.separator}`);
  console.log(M.obj1_title);
  console.log(M.obj1_desc);
  console.log(M.separator);

  const checkpoints = [100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000]
    .filter(cp => cp <= N);

  const results = [];
  let prevBN = NaN;

  console.log(`\n${padR(M.obj1_n, 10)}  ${padR(M.obj1_bN, 14)}  ${padR(M.obj1_status, 8)}`);
  console.log(`${padR('──────────', 10)}  ${padR('──────────────', 14)}  ${padR('────────', 8)}`);

  for (const cp of checkpoints) {
    const idx = Array.from({ length: cp }, (_, i) => i + 1);
    const gram = gramPointsVec(idx);
    let sum = 0;
    for (let i = 0; i < cp; i++) sum += Math.abs(zeros[i] - gram[i]);
    const bN = sum / cp;

    let status;
    if (isNaN(prevBN)) status = '—';
    else if (bN < prevBN * 1.05) status = '↓';
    else status = '↑';

    console.log(`${padL(cp, 10)}  ${padL(bN.toFixed(8), 14)}  ${padR(status, 8)}`);
    results.push({ N: cp, bN, status });
    prevBN = bN;
  }

  const finalBN = results[results.length - 1].bN;
  let verdict;
  if (finalBN < 0.01) verdict = M.obj1_converge;
  else if (finalBN < 0.5) verdict = M.obj1_stable;
  else verdict = M.obj1_diverge;

  console.log(`\n${verdict}`);
  return results;
}

// ==============================================================================
// Objection 2: GUE Spacing KS Test
// ==============================================================================

function objection2(zeros, M) {
  const N = zeros.length;
  console.log(`\n${M.separator}`);
  console.log(M.obj2_title);
  console.log(M.obj2_desc);
  console.log(M.separator);

  // Compute normalized spacings
  const spacings = [];
  for (let k = 0; k < N - 1; k++) {
    const delta = zeros[k + 1] - zeros[k];
    const logFac = Math.log(zeros[k] / (2 * Math.PI)) / (2 * Math.PI);
    const s = delta * logFac;
    if (isFinite(s) && s > 0) spacings.push(s);
  }

  // Sort for empirical CDF
  spacings.sort((a, b) => a - b);
  const n = spacings.length;

  // KS test: max |F_emp(s) - F_GUE(s)|
  let Dplus = 0, Dminus = 0;
  for (let i = 0; i < n; i++) {
    const empVal = (i + 1) / n;
    const empPrev = i / n;
    const gueVal = gueCDF(spacings[i]);
    Dplus  = Math.max(Dplus, empVal - gueVal);
    Dminus = Math.max(Dminus, gueVal - empPrev);
  }
  const D = Math.max(Dplus, Dminus);

  // Kolmogorov p-value approximation
  const lambda = (Math.sqrt(n) + 0.12 + 0.11 / Math.sqrt(n)) * D;
  let pVal = 0;
  for (let k = -5; k <= 5; k++) {
    pVal += Math.pow(-1, k) * Math.exp(-2 * k * k * lambda * lambda);
  }
  pVal = Math.max(0, Math.min(1, pVal));

  console.log(`\n${padR(M.obj2_stat + ':', 18)} ${D.toFixed(8)}`);
  console.log(`${padR(M.obj2_pval + ':', 18)} ${pVal.toExponential(6)}`);

  const verdict = pVal > 0.05 ? M.obj2_pass : M.obj2_fail;
  console.log(`\n${verdict}`);
  return { D, pValue: pVal };
}

// ==============================================================================
// Objection 3: Large-T Decay Slope
// ==============================================================================

function objection3(zeros, M) {
  const N = zeros.length;
  console.log(`\n${M.separator}`);
  console.log(M.obj3_title);
  console.log(M.obj3_desc);
  console.log(M.separator);

  // Use upper half of data (large T region)
  const start = Math.max(1, Math.floor(N * 0.5));
  const idx = [];
  for (let i = start; i <= N; i++) idx.push(i);

  const gram = gramPointsVec(idx);
  const logGamma = [];
  const logDev = [];

  for (let i = 0; i < idx.length; i++) {
    const dev = Math.abs(zeros[idx[i] - 1] - gram[i]);
    if (isFinite(dev) && dev > 0) {
      logGamma.push(Math.log(zeros[idx[i] - 1]));
      logDev.push(Math.log(dev));
    }
  }

  // Linear regression: logDev = a + b * logGamma
  const n = logGamma.length;
  let sumX = 0, sumY = 0, sumXY = 0, sumX2 = 0;
  for (let i = 0; i < n; i++) {
    sumX  += logGamma[i];
    sumY  += logDev[i];
    sumXY += logGamma[i] * logDev[i];
    sumX2 += logGamma[i] * logGamma[i];
  }
  const denom = n * sumX2 - sumX * sumX;
  const intercept = (sumY * sumX2 - sumX * sumXY) / denom;
  const slope = (n * sumXY - sumX * sumY) / denom;

  // Standard error of slope
  let ssRes = 0;
  for (let i = 0; i < n; i++) {
    ssRes += Math.pow(logDev[i] - (intercept + slope * logGamma[i]), 2);
  }
  const s2 = ssRes / (n - 2);
  const stderr = Math.sqrt(s2 * n / denom);

  console.log(`\n${padR(M.obj3_slope + ':', 18)} ${slope.toFixed(6)}`);
  console.log(`${padR(M.obj3_stderr + ':', 18)} ${stderr.toFixed(6)}`);
  console.log(`${padR(M.obj3_target + ':', 18)} -0.5`);

  const verdict = Math.abs(slope - (-0.5)) < 0.15 ? M.obj3_pass : M.obj3_fail;
  console.log(`\n${verdict}`);
  return { slope, stderr };
}

// ==============================================================================
// String padding helpers
// ==============================================================================

function padL(str, len) { return String(str).padStart(len); }
function padR(str, len) { return String(str).padEnd(len); }

// ==============================================================================
// Main verification function
// ==============================================================================

async function abCloudVerify(options = {}) {
  const {
    dataDir   = '../data',
    zeros     = 0,
    source    = 'auto',
    objection = 'all',
    lang      = 'en',
  } = options;

  const M = MSG[lang] || MSG.en;

  console.log(`\n${M.separator}`);
  console.log(M.header);
  console.log(M.separator);

  // Load zeros
  const gammas = await loadZeros(dataDir, zeros, source);
  if (gammas.length === 0) {
    console.log(M.no_data);
    return null;
  }

  // Run selected objections
  const results = {};
  if (['all', '1'].includes(objection)) results.obj1 = objection1(gammas, M);
  if (['all', '2'].includes(objection)) results.obj2 = objection2(gammas, M);
  if (['all', '3'].includes(objection)) results.obj3 = objection3(gammas, M);

  console.log(`\n${M.separator}`);
  console.log(M.done);
  console.log('');
  return results;
}

// ==============================================================================
// Exports
// ==============================================================================

module.exports = {
  abCloudVerify,
  loadZeros,
  objection1,
  objection2,
  objection3,
  lambertW0,
  gramPoint,
  gramPointsVec,
  gueCDF,
  MSG,
};

File: ab-cloud-verification/julia/ab_cloud_verify.jl
#!/usr/bin/env julia
# =============================================================================
# AB-CLOUD VERIFICATION SUITE — BILINGUAL (EN/RU)
# Idiomatic Julia · Type Annotations · Efficient
# =============================================================================
# Проверочный пакет AB-Cloud — билингвальный (английский/русский)
# Три возражения против гипотезы Римана:
#   Возражение 1: b(N) → 0?  (сходимость отклонений Грама)
#   Возражение 2: GUE间距 KS-тест (уровневые промежутки)
#   Возражение 3: Large-T decay slope ≈ -0.5
# =============================================================================

using Printf

# =============================================================================
# Constants
# =============================================================================
const PI::Float64     = 3.14159265358979323846
const TWO_PI::Float64 = 6.28318530717958647693

# =============================================================================
# Data structures
# =============================================================================
struct VerifyResult
    label::String
    value::Float64
    pass::Bool
    detail::String
end

# =============================================================================
# Load zeta zeros with auto-selection
# =============================================================================
function load_zeros(n_request::Int; source::String="auto")::Vector{Float64}
    filepath::String = if source != "auto"
        "../data/$source"
    elseif n_request ≤ 13661
        "../data/zeta_zeros_50000.txt"
    elseif n_request ≤ 500_000
        "../data/zeta_zeros_500k_odlyzko.txt"
    elseif n_request ≤ 2_000_000
        "../data/zeta_zeros_2M_odlyzko.txt"
    else
        "../data/zeros6.txt"
    end

    println("  Data file: $filepath")
    zeros::Vector{Float64} = Float64[]

    open(filepath, "r") do io
        for line in eachline(io)
            line = strip(line)
            isempty(line) && continue
            startswith(line, '#') && continue
            try
                v = parse(Float64, line)
                v > 0.0 && push!(zeros, v)
            catch
                continue
            end
        end
    end

    if length(zeros) > n_request
        resize!(zeros, n_request)
    end

    return zeros
end

# =============================================================================
# Lambert W (principal branch) via Halley's method
# =============================================================================
function lambert_w0(x::Float64)::Float64
    x == 0.0 && return 0.0
    w::Float64 = log(max(x, 1e-30))
    w > 0.0 && (w = log(w))

    for _ in 1:50
        ew::Float64  = exp(w)
        f::Float64   = w * ew - x
        fp::Float64  = ew * (w + 1.0)
        fpp::Float64 = ew * (w + 2.0)
        delta::Float64 = f / (fp - 0.5 * f * fpp / fp)
        w -= delta
        abs(delta) < 1e-15 * abs(w) && break
    end
    return w
end

# =============================================================================
# Gram point: γ̃_n = 2π · W₀(n/e)
# =============================================================================
gram_point(n::Int)::Float64 = TWO_PI * lambert_w0(n / exp(1.0))

# =============================================================================
# Objection 1: b(N) = (1/N) * Σ|γ_k - γ̃_k|
# =============================================================================
function objection1(gammas::Vector{Float64}; lang::Int=0)::VerifyResult
    N::Int = length(gammas)
    s::Float64 = 0.0
    @inbounds for k in 1:N
        s += abs(gammas[k] - gram_point(k))
    end
    b_n::Float64 = s / N
    pass::Bool = b_n < 1.0

    if lang == 2
        println("\n═══════════════════════════════════════════════════")
        println("  ВОЗРАЖЕНИЕ 1: Сходимость b(N)")
        println("  b(N) = (1/N) · Σ|γ_k − γ̃_k|, точки Грама через W Ламберта")
        println("═══════════════════════════════════════════════════")
        @printf("    N       = %d\n", N)
        @printf("    b(N)    = %.8e\n", b_n)
        println("    Пройдено: $pass")
    else
        println("\n═══════════════════════════════════════════════════")
        println("  OBJECTION 1: b(N) Convergence")
        println("  b(N) = (1/N) · Σ|γ_k − γ̃_k|, Gram pts via Lambert W")
        println("═══════════════════════════════════════════════════")
        @printf("    N       = %d\n", N)
        @printf("    b(N)    = %.8e\n", b_n)
        println("    Pass:   $pass")
    end

    return VerifyResult("b(N)", b_n, pass, @sprintf("%.8e", b_n))
end

# =============================================================================
# GUE PDF: p(s) = (πs/2) · exp(-πs²/4)
# =============================================================================
gue_pdf(s::Float64)::Float64 = (PI * s / 2.0) * exp(-PI * s^2 / 4.0)

# =============================================================================
# GUE CDF via Simpson's rule
# =============================================================================
function gue_cdf(s::Float64)::Float64
    s ≤ 0.0 && return 0.0
    nsteps::Int = 200
    h::Float64 = s / nsteps
    sum_val::Float64 = 0.0
    @inbounds for i in 0:nsteps
        x = i * h
        if i == 0 || i == nsteps
            sum_val += gue_pdf(x)
        elseif isodd(i)
            sum_val += 4.0 * gue_pdf(x)
        else
            sum_val += 2.0 * gue_pdf(x)
        end
    end
    return min(h / 3.0 * sum_val, 1.0)
end

# =============================================================================
# Approximate Kolmogorov p-value
# =============================================================================
function kolmogorov_pvalue(d::Float64, n::Float64)::Float64
    z::Float64 = d * sqrt(n)
    pval::Float64 = 0.0
    for k in -10:10
        term::Float64 = exp(-2.0 * (2k * z + z)^2)
        pval += (-1)^k * term
    end
    return clamp(pval, 0.0, 1.0)
end

# =============================================================================
# Objection 2: GUE spacing KS test
# s_k = (γ_{k+1} - γ_k) · log(γ_k/(2π)) / (2π)
# =============================================================================
function objection2(gammas::Vector{Float64}; lang::Int=0)::VerifyResult
    N::Int = length(gammas)
    m::Int = N - 1
    s::Vector{Float64} = Vector{Float64}(undef, m)

    @inbounds for k in 1:m
        log_factor::Float64 = log(gammas[k] / TWO_PI)
        s[k] = (gammas[k+1] - gammas[k]) * log_factor / TWO_PI
    end

    # Sort for empirical CDF
    sort!(s)

    # KS statistic
    ks_stat::Float64 = 0.0
    @inbounds for k in 1:m
        cdf_emp::Float64  = k / m
        cdf_theo::Float64 = gue_cdf(s[k])
        d_plus::Float64   = abs(cdf_emp - cdf_theo)
        d_plus > ks_stat && (ks_stat = d_plus)
    end

    ks_pval::Float64 = kolmogorov_pvalue(ks_stat, Float64(m))
    pass::Bool = ks_pval > 0.05

    if lang == 2
        println("\n═══════════════════════════════════════════════════")
        println("  ВОЗРАЖЕНИЕ 2: GUE-интервалы, KS-критерий")
        println("  s_k = Δγ_k · log(γ_k/2π) / 2π,  p(s) = (πs/2)·e^{-πs²/4}")
        println("═══════════════════════════════════════════════════")
        @printf("    Кол-во интервалов = %d\n", m)
        @printf("    KS-статистика     = %.8e\n", ks_stat)
        @printf("    p-значение        = %.8e\n", ks_pval)
        println("    Пройдено (p>0.05): $pass")
    else
        println("\n═══════════════════════════════════════════════════")
        println("  OBJECTION 2: GUE Spacing KS Test")
        println("  s_k = Δγ_k · log(γ_k/2π) / 2π,  p(s) = (πs/2)·e^{-πs²/4}")
        println("═══════════════════════════════════════════════════")
        @printf("    Spacing count    = %d\n", m)
        @printf("    KS statistic     = %.8e\n", ks_stat)
        @printf("    p-value          = %.8e\n", ks_pval)
        println("    Pass (p>0.05):   $pass")
    end

    return VerifyResult("KS", ks_stat, pass, @sprintf("p=%.4e", ks_pval))
end

# =============================================================================
# Objection 3: Large-T decay slope ≈ -0.5
# Fit log|Δγ_k| vs log(γ_k)
# =============================================================================
function objection3(gammas::Vector{Float64}; lang::Int=0)::VerifyResult
    N::Int = length(gammas)
    i_start::Int = div(N, 2) + 1
    m::Int = N - i_start

    if m < 10
        return VerifyResult("slope", 0.0, false, "insufficient data")
    end

    # Compute means
    x_mean::Float64 = 0.0
    y_mean::Float64 = 0.0
    @inbounds for k in i_start:(N-1)
        x_mean += log(gammas[k])
        y_mean += log(abs(gammas[k+1] - gammas[k]))
    end
    x_mean /= m
    y_mean /= m

    # Linear regression
    sxx::Float64 = 0.0
    sxy::Float64 = 0.0
    @inbounds for k in i_start:(N-1)
        xv::Float64 = log(gammas[k])
        yv::Float64 = log(abs(gammas[k+1] - gammas[k]))
        sxx += (xv - x_mean)^2
        sxy += (xv - x_mean) * (yv - y_mean)
    end
    slope::Float64 = sxy / sxx
    deviation::Float64 = abs(slope + 0.5)
    pass::Bool = deviation < 0.15

    if lang == 2
        println("\n═══════════════════════════════════════════════════")
        println("  ВОЗРАЖЕНИЕ 3: Наклон убывания при больших T")
        println("  log|Δγ_k| ~ slope · log(γ_k),  ожидается slope ≈ -0.5")
        println("═══════════════════════════════════════════════════")
        @printf("    Точек регрессии = %d\n", m)
        @printf("    Наклон (slope)  = %.4f\n", slope)
        @printf("    Отклонение      = %.4f\n", deviation)
        println("    Пройдено:       $pass")
    else
        println("\n═══════════════════════════════════════════════════")
        println("  OBJECTION 3: Large-T Decay Slope")
        println("  log|Δγ_k| ~ slope · log(γ_k),  expected slope ≈ -0.5")
        println("═══════════════════════════════════════════════════")
        @printf("    Regression pts  = %d\n", m)
        @printf("    Slope           = %.4f\n", slope)
        @printf("    Deviation       = %.4f\n", deviation)
        println("    Pass:           $pass")
    end

    return VerifyResult("slope", slope, pass, @sprintf("dev=%.4f", deviation))
end

# =============================================================================
# Summary table
# =============================================================================
function print_summary(results::Vector{VerifyResult}; lang::Int=0)
    if lang == 2
        println("\n  ┌──────────┬────────────────────┬────────┐")
        println("  │ Возраж.  │    Значение        │ Стат.  │")
        println("  ├──────────┼────────────────────┼────────┤")
    else
        println("\n  ┌──────────┬────────────────────┬────────┐")
        println("  │ Obj.     │    Value           │ Status │")
        println("  ├──────────┼────────────────────┼────────┤")
    end

    labels = lang == 2 ? ["1:b(N)", "2:KS", "3:slope"] : ["1:b(N)", "2:KS", "3:slope"]
    for (i, r) in enumerate(results)
        status = r.pass ? "PASS" : "FAIL"
        @printf("  │ %-8s │ %16.6e │ %-6s │\n", labels[i], r.value, status)
    end

    println("  └──────────┴────────────────────┴────────┘")
end

# =============================================================================
# Banner
# =============================================================================
function print_banner(; lang::Int=0)
    println()
    if lang == 2
        println("  ╔═══════════════════════════════════════════════╗")
        println("  ║   AB-CLOUD ПРОВЕРКА — ГИПОТЕЗА РИМАНА        ║")
        println("  ║   Три возражения: b(N), GUE KS, Large-T      ║")
        println("  ╚═══════════════════════════════════════════════╝")
    else
        println("  ╔═══════════════════════════════════════════════╗")
        println("  ║   AB-CLOUD VERIFICATION — RIEMANN HYPOTHESIS  ║")
        println("  ║   Three objections: b(N), GUE KS, Large-T     ║")
        println("  ╚═══════════════════════════════════════════════╝")
    end
    println()
end

# =============================================================================
# Main entry point
# =============================================================================
function main()
    # Defaults
    n_zeros::Int    = 10000
    source::String  = "auto"
    objection::Int  = 0   # 0 = all
    lang::Int       = 0   # 0 = bilingual

    # Parse args
    args = ARGS
    i = 1
    while i ≤ length(args)
        arg = args[i]
        if arg == "--zeros" && i+1 ≤ length(args)
            i += 1; n_zeros = parse(Int, args[i])
        elseif arg == "--source" && i+1 ≤ length(args)
            i += 1; source = args[i]
        elseif arg == "--objection" && i+1 ≤ length(args)
            i += 1
            objection = args[i] == "all" ? 0 : parse(Int, args[i])
        elseif arg == "--lang" && i+1 ≤ length(args)
            i += 1
            lang = args[i] == "en" ? 1 : (args[i] == "ru" ? 2 : 0)
        end
        i += 1
    end

    print_banner(lang=lang)

    gammas::Vector{Float64} = load_zeros(n_zeros; source=source)
    N::Int = length(gammas)

    if lang == 2
        println("  Загружено нулей: $N (γ-ординаты)")
    else
        println("  Zeros loaded: $N (gamma ordinates)")
    end

    results::Vector{VerifyResult} = VerifyResult[]

    t_start = time()

    if objection == 0 || objection == 1
        push!(results, objection1(gammas; lang=lang))
    end
    if objection == 0 || objection == 2
        push!(results, objection2(gammas; lang=lang))
    end
    if objection == 0 || objection == 3
        push!(results, objection3(gammas; lang=lang))
    end

    print_summary(results; lang=lang)

    t_end = time()
    if lang == 2
        @printf("  Общее время: %.3f с\n", t_end - t_start)
    else
        @printf("  Total time: %.3f s\n", t_end - t_start)
    end
end

main()
